### PR TITLE
[ESI] Start of system manifest: types

### PR DESCRIPTION
**Pull request by teqdruid**
An ESI system manifest is intended to describe the parts of a particular accelerator which may be relevant to software. A compressed version of it will be embedded in accelerators. It will eventually replace the services metadata json.

This first PR only outputs types on cosim ports.

### CODE REVIEW

File: include/circt/Dialect/ESI/ESIPasses.td [hunk 1]
```diff
@@ -51,6 +51,16 @@ def ESIEmitCollateral: Pass<"esi-emit-collateral", "mlir::ModuleOp"> {
   ];
 }
 
+def ESIBuildManifest : Pass<"esi-build-manifest", "mlir::ModuleOp"> {
+  let summary = "Build a manifest of an ESI system";
+  let constructor = "circt::esi::createESIBuildManifestPass()";
+  let dependentDialects = ["circt::hw::HWDialect", "circt::sv::SVDialect"];
+  let options = [
+    Option<"writeDirect", "write-direct", "std::string",
+           "", "Write the manifest JSON directly to this file">
```
Comment (mortbopet) [inline include/circt/Dialect/ESI/ESIPasses.td:60]: funny option name - how about `dump` or `to-file`, ...?

**Review by mortbopet — COMMENTED** [review]
Do we want to go directly from inspecting the IR to dumping the manifest to JSON? In my mind, there would be a pass which constructs the manifest and attaches it to the IR in some easily digestible format (... maybe that _is_ just a json string, but perhaps not embedded inside an sv.verbatim op already). The reason for this being that it allows other passes to depend on the manifest for lookup. One may then say that such passes should just directly inspect the IR, but i think the value of the manifest (and this pass) is that it contains all of the logic to inspect the design and put the relevant information into a properly structured format, which we don't want to duplicate in other passes.

Other than that, the implementation LGTM.

File: include/circt/Dialect/ESI/ESIPasses.td (hunk 1 above)
Comment (teqdruid) [reply include/circt/Dialect/ESI/ESIPasses.td]: changed

**Comment by teqdruid** [comment]
That's how I started, but couldn't come up with a compelling reason to do that just for types. I will have ops for other types of metadata and appid stuff which this pass will consume and output to json.

**Pull request merged**